perception/autoware_tensorrt_bevdet/src/bevdet.cpp

```diff
@@ -0,0 +1,687 @@
+  y_end_ = model_config["grid_config"]["y"][1].as<float>();
+  y_step_ = model_config["grid_config"]["y"][2].as<float>();
+  z_start_ = model_config["grid_config"]["z"][0].as<float>();
+  z_end_ = model_config["grid_config"]["z"][1].as<float>();
+  z_step_ = model_config["grid_config"]["z"][2].as<float>();
+  bevpool_channel_ = model_config["model"]["bevpool_channels"].as<int>();
+  nms_pre_maxnum_ = model_config["test_cfg"]["max_per_img"].as<int>();
+  nms_post_maxnum_ = model_config["test_cfg"]["post_max_size"].as<int>();
+  score_thresh_ = model_config["test_cfg"]["score_threshold"].as<float>();
+  nms_overlap_thresh_ = model_config["test_cfg"]["nms_thr"][0].as<float>();
+  use_depth_ = model_config["use_depth"].as<bool>();
+  use_adj_ = model_config["use_adj"].as<bool>();
+  transform_size_ = model_config["transform_size"].as<int>();
+  cam_params_size_ = model_config["cam_params_size"].as<int>();
+
+  std::vector<std::vector<float>> nms_factor_temp =
+    model_config["test_cfg"]["nms_rescale_factor"].as<std::vector<std::vector<float>>>();
+  nms_rescale_factor_.clear();
+  for (const auto & task_factors : nms_factor_temp) {
+    for (float factor : task_factors) {
+      nms_rescale_factor_.push_back(factor);
+    }
+  }
+
+  std::vector<std::vector<std::string>> class_name_pre_task;
+  class_num_ = 0;
+  YAML::Node tasks = model_config["model"]["tasks"];
+  class_num_pre_task_ = std::vector<int>();
+  for (auto it : tasks) {
+    int num = it["num_class"].as<int>();
+    class_num_pre_task_.push_back(num);
+    class_num_ += num;
+    class_name_pre_task.push_back(it["class_names"].as<std::vector<std::string>>());
+  }
+
+  YAML::Node common_head_channel = model_config["model"]["common_head"]["channels"];
+  YAML::Node common_head_name = model_config["model"]["common_head"]["names"];
+  for (size_t i = 0; i < common_head_channel.size(); i++) {
+    out_num_task_head_[common_head_name[i].as<std::string>()] = common_head_channel[i].as<int>();
+  }
+
+  feat_h_ = input_img_h_ / down_sample_;
+  feat_w_ = input_img_w_ / down_sample_;
+  depth_num_ = (depth_end_ - depth_start_) / depth_step_;
+  xgrid_num_ = (x_end_ - x_start_) / x_step_;
+  ygrid_num_ = (y_end_ - y_start_) / y_step_;
+  zgrid_num_ = (z_end_ - z_start_) / z_step_;
+  bev_h_ = ygrid_num_;
+  bev_w_ = xgrid_num_;
+
+  post_rot_ << resize_radio_, 0, 0, 0, resize_radio_, 0, 0, 0, 1;
+  post_trans_.translation() << -crop_w_, -crop_h_, 0;
+
+  adj_num_ = 0;
+  if (use_adj_) {
+    adj_num_ = model_config["adj_num"].as<int>();
+  }
+
+  postprocess_ptr_.reset(new PostprocessGPU(
+    class_num_, score_thresh_, nms_overlap_thresh_, nms_pre_maxnum_, nms_post_maxnum_, down_sample_,
+    bev_h_, bev_w_, x_step_, y_step_, x_start_, y_start_, class_num_pre_task_,
+    nms_rescale_factor_));
+}
+
+void BEVDet::mallocDeviceMemory()
+{
+  trt_buffer_sizes_.resize(trt_engine_->getNbBindings());
+  trt_buffer_dev_ = reinterpret_cast<void **>(new void *[trt_engine_->getNbBindings()]);
```

Comment:
@cyn-liu 

Please apply the changes not only to the point we pointed out, but to all points related to the comment.

As we said in https://github.com/autowarefoundation/autoware.universe/pull/7956#discussion_r1743792431, don't use new.
Please check our coding guidelines:

https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/
Our coding guideline follows [CppCoreGuidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#r11-avoid-calling-new-and-delete-explicitly)

Reply:
Refactoring the entire open source project to fully meet the specifications is a significant workload, but I have to do so.
 I found that the refactored code often encounters unknown problems during launch, and the cause of this problem has not been found. After restarting the node multiple times, it can be resolved. Although the previous code did not meet your coding standards, it was simple clear, and can ran stably.

Reply:
@cyn-liu 
Is this what you meant in your comment?

- `autoware_tensorrt_bevdet/src/bevdet.cpp` was copied from another open source project.
- The content written in the other open source project is being used as is.

If that's the case, instead of modifying the code, the following actions are also possible:

- Clearly indicate where the code was copied from, like in [this example](https://github.com/autowarefoundation/autoware.universe/tree/main/perception/autoware_bytetrack#license).
- Store all copied code in a `lib` directory.

For code outside the `lib`, you have to follow Autoware's coding guidelines. For example, Autoware’s coding guidelines recommend avoiding the use of `new`.

Reply:
@cyn-liu (cc. @Shin-kyoto)
With all due respect, but

> Although the previous code did not meet your coding standards

All PR authors are asked to read and accept the coding standards in the contribution guidelines, so this is a requirement that was clear even before the work on this PR started, so we expect this to be followed. A large open source project cannot be managed otherwise and this is not a concept unique to Autoware.

> it was simple clear

I disagree, and so do our coding standards and the C++ core guidelines. Also, this PR is almost 6000 lines which makes even perfectly simple and clear code hard to review thoroughly.

> and can ran stably.

As a reviewer, this was impossible to verify due to all the non-standard and almost C-style C++. There is a reason that core guidelines and tools like Clang-Tidy are enforcing the use of smart pointers, disallowing reinterpret_cast etc.
If it does not run stably after the refactoring, that hints at race conditions and memory problems that still need to be addressed. Consider using AddressSanitizer and ThreadSanitizer to pinpoint the issues.


As this review is taking quite some hours to review for our employees, I would strongly suggest that the coding style issues are addressed, and that the PR is split up before occupying people with the same discussion over and over again.

If you do not want to comply with our coding standards, I am afraid we cannot merge this PR.

Reply:
@cyn-liu 

friendly ping

Reply:
@cyn-liu 

friendly ping

Reply:
@cyn-liu 

I have talked with @Shin-kyoto on this.Like @mojomex mentioned, we have a coding standard in AWF which encourages the use of smart pointers instead of normal pointers. Therefore, we would like to comply to the guideline as much as possible. 

However, I also understand the effort of refactoring the whole code, especially if the code was taken from another open source. If it is difficult to refactor, alternate option is to wrap the original BEVDet code as is as a vendor package and maintain it outside Autoware with different CIs settings and code standards. Then we can added through autoware.repos or through ROS Buildfarm and just call the libraries from Autoware Nodes. This might also make it easier for us to update the BEVDet implementation if any updates are made upstream.

Do you think the above approach is feasible? It might be difficult if you have already made a lot of modification to the BEVDet implementation so it would be nice if you can provide us how much modification you made from the [original code](https://github.com/HuangJunJie2017/BEVDet).

I'm also okay to set up an online call if that makes you easier to discuss on this topic.

Reply:
Good Afternoon, Mits-san @mitsudome-r  . thank you for the coordination.

1. Second way is a better option for Continuously introduce great open source project into autoware
2. I will tell Cynthia @cyn-liu  to make a list of the modification point here
3. it will be great if we can have a online call for this， especially for the topic “call the libraries from Autoware Nodes”， we are not familiar with this part

have a nice day!
Lucas

Reply:
The main modifications made to `tensorrt-bevdet-cpp` library are as follows:
1. Standardize log output by changing `print` ,`std:: cout` etc to ros2 `rclcpp logger`
2. Standardize the names of member functions and variables
3. Modify multiple pointers to cpp `std::vector`
4. If the engine model does not exist,export engine before inference